chia/wallet/puzzles/nft_state_layer.clvm

```diff
@@ -28,62 +34,55 @@
+              )
+              (c (f conditions) (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (r conditions) odd_coin_params new_metadata_info metadata_seen))
+          )
+          (if (= (f (f conditions)) -24)
+              (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (r conditions) odd_coin_params
+                (assert (all
+                          (= (sha256tree1 (f (r (f conditions)))) (f (r (f new_metadata_info))))
+                          (not metadata_seen)
+                        )
+                    ; then
+                    (a (f (r (f conditions))) (list (f (f new_metadata_info)) (f (r (f new_metadata_info))) (f (r (r (f conditions))))))
+                )
+                1  ; the metadata update has been seen now
+              )
+              (c (f conditions) (wrap_odd_create_coins NFT_STATE_LAYER_MOD_HASH (r conditions) odd_coin_params new_metadata_info metadata_seen))
+          )
       )
-      ()
-    )
-  )
-
-  ; take two lists and merge them into one
-  (defun merge_list (list_a list_b)
-    (if list_a
-      (c (f list_a) (merge_list (r list_a) list_b))
-      list_b
-    )
-  )
-
-  (defun check_for_metadata_prog_reveal (METADATA_UPDATER_PUZZLE_HASH CURRENT_METADATA conditions_loop)
-    (if conditions_loop
-      (if (= (f (f conditions_loop)) -24)
-        (if (= (sha256tree1 (f (r (f conditions_loop)))) METADATA_UPDATER_PUZZLE_HASH)
-          (a (f (r (f conditions_loop))) (list CURRENT_METADATA METADATA_UPDATER_PUZZLE_HASH (f (r (r (f conditions_loop))))))
-          (x)
+      (c
+        (c CREATE_COIN
+            (c
+              (nft_state_layer_puzzle_hash
+                NFT_STATE_LAYER_MOD_HASH
+                (f (f new_metadata_info))
+                (f (r (f new_metadata_info)))
+                (f odd_coin_params)  ; metadata updater solution
+              )
+              (r odd_coin_params)
+            )
         )
-        (check_for_metadata_prog_reveal METADATA_UPDATER_PUZZLE_HASH CURRENT_METADATA (r conditions_loop))
+        (f (r new_metadata_info))  ; metadata_updater conditions
       )
-      (list (list CURRENT_METADATA METADATA_UPDATER_PUZZLE_HASH) 0)
     )
   )
 
-  (defun process_metadata_updater (metadata_result conditions)
-    (c (f metadata_result) (list (merge_list (f (r metadata_result)) conditions)))
-  )
-
-  (defun metadata_updater_loader (METADATA_UPDATER_PUZZLE_HASH CURRENT_METADATA conditions)
-    (process_metadata_updater (check_for_metadata_prog_reveal METADATA_UPDATER_PUZZLE_HASH CURRENT_METADATA conditions) conditions)
-  )
-
   ; main
-  (c
-    (list ASSERT_MY_AMOUNT my_amount)
-    (wrap_odd_create_coins
-      NFT_STATE_LAYER_MOD_HASH
-      (metadata_updater_loader
-       METADATA_UPDATER_PUZZLE_HASH METADATA (a INNER_PUZZLE solution)
-      )
-      my_amount
-      0
-    )
+  (wrap_odd_create_coins
+    NFT_STATE_LAYER_MOD_HASH
+    (a INNER_PUZZLE inner_solution)
+    ()
```

Comment:
Why hardcode () here?

Reply:
We start the loop with this set to `()` and we expect it to get set during the loop.  We check at the end of the recursion whether or not it's still `()`, and if it is, we raise.